sycl/doc/design/SYCLNativeCPU.md

```diff
@@ -0,0 +1,172 @@
+  -G Ninja \
+  -B build -S . \
+ -DCMAKE_CXX_COMPILER=clang++ \
+ -DSYCL_TEST_E2E_TARGETS="native_cpu:cpu" 
+
+```
+
+Note that a number of `e2e` tests are currently still failing.
+
+# Running example
+
+The following section gives a brief overview of how a simple SYCL application is compiled for the Native CPU target. Consider the following SYCL sample, which performs vector addition using USM:
+
+```c++
+  cl::sycl::queue deviceQueue;
+  cl::sycl::range<1> numOfItems{N};
+  auto a_ptr = sycl::malloc_device<int>(N, deviceQueue);
+  auto b_ptr = sycl::malloc_device<int>(N, deviceQueue);
+  auto c_ptr = sycl::malloc_device<int>(N, deviceQueue);
+
+  // copy mem to device, omitted
+  deviceQueue
+      .submit([&](cl::sycl::handler &cgh) {
+        auto kern = [=](cl::sycl::id<1> wiID) {
+          c_ptr[wiID] = a_ptr[wiID] + b_ptr[wiID];
+        };
+        cgh.parallel_for<class Sample>(numOfItems, kern);
+      })
+      .wait();
+  deviceQueue.memcpy(C.data(), c_ptr, N * sizeof(int));
+
+```
+
+The extracted device code produces the following LLVM-IR:
+
+```llvm
+define weak_odr dso_local spir_kernel void @_Z6Sample(ptr noundef align 4 %_arg_c_ptr, ptr noundef align 4 %_arg_a_ptr, ptr noundef align 4 %_arg_b_ptr) local_unnamed_addr #1 comdat !srcloc !74 !kernel_arg_buffer_location !75 !kernel_arg_type !76 !sycl_fixed_targets !49 !sycl_kernel_omit_args !77 {
+entry:
+  %0 = load i64, ptr @__spirv_BuiltInGlobalInvocationId, align 32, !noalias !78
+  %arrayidx.i = getelementptr inbounds i32, ptr %_arg_a_ptr, i64 %0
+  %1 = load i32, ptr %arrayidx.i, align 4, !tbaa !72
+  %arrayidx4.i = getelementptr inbounds i32, ptr %_arg_b_ptr, i64 %0
+  %2 = load i32, ptr %arrayidx4.i, align 4, !tbaa !72
+  %add.i = add nsw i32 %1, %2
+  %cmp.i8.i = icmp ult i64 %0, 2147483648
+  tail call void @llvm.assume(i1 %cmp.i8.i)
+  %arrayidx6.i = getelementptr inbounds i32, ptr %_arg_c_ptr, i64 %0
+  store i32 %add.i, ptr %arrayidx6.i, align 4, !tbaa !72
+  ret void
+}
+```
+
+For the Native CPU target, the device compiler needs to perform two main operations:
+* Materialize the SPIRV builtins (such as `@__spirv_BuiltInGlobalInvocationId`), so that they can be correctly updated by the runtime when executing the kernel. This is performed by the [PrepareSYCLNativeCPU pass](llvm/lib/SYCLLowerIR/PrepareSYCLNativeCPU.cpp).
+* Allow the SYCL runtime to call the kernel, by registering it to the SYCL runtime, operation performed by the [EmitSYCLNativeCPUHeader pass](llvm/lib/SYCLLowerIR/EmitSYCLNativeCPUHeader.cpp).
+
+
+## PrepareSYCLNativeCPU Pass
+
+This pass will add a pointer to a `nativecpu_state` struct as kernel argument to all the kernel functions, and it will replace all the uses of SPIRV builtins with the return value of appropriately defined functions, which will read the requested information from the `nativecpu_state` struct. The `nativecpu_state` struct and the builtin functions are defined in [native_cpu.hpp](sycl/include/sycl/detail/native_cpu.hpp).
```

Comment:
Could you please confirm that the state argument is added to all device functions accessible from any kernel?
If so, it can cause problems with debuggers, e.g. a user may try to call a function which originally accepts no arguments, well, without arguments, but the function itself will expect that the `state` arg is passed and possibly use incorrect memory.

Have you considered using `thread_local` variable for the state instead of a function argument? This could simplify the implementation and reduce required transformation.

Reply:
You are right, our plan is to add the state argument to all the device functions that need it (currently we add it just to the kernel and rely on compiler optimizations, but we were planning on doing it). I've never considered this concern with debuggers, thanks for bringing it up, using `thread_local` variables sounds like a good solution, and @AlexeySachkov said it's probably better to implement it in a following PR, thank you